documenting get_moment (#5244)
pymc-devs/pymc: docs/source/contributing/developer_guide_implementing_distribution.md

It is designed for developers who wish to add a new distribution to the library.
Users will not be aware of all this complexity and should instead make use of helper methods such as (TODO).

> PyMC {class}`~pymc.distributions.Distribution` build on top of Aesara's {class}`~aesara.tensor.random.op.RandomVariable`, and implement `logp`, `logcdf` and `get_moment` methods as well as other initialization and validation helpers, most notably `shape/dims`, alternative parametrizations, and default `transforms`.

Reviewer: ```suggestion
PyMC {class}`~pymc.distributions.Distribution` builds on top of Aesara's {class}`~aesara.tensor.random.op.RandomVariable`, and implements `logp`, `logcdf` and `get_moment` methods as well as other initialization and validation helpers.
Most notably `shape/dims` kwargs, alternative parametrizations, and default `transforms`.
```

        # the rv_op needs in order to be instantiated
        return super().dist([param1, param2], **kwargs)

    # get_moment returns the stable moment from which to start sampling for the

>     # distribution given the implicit `rv`, `size` and `param1` ... `paramN`

Reviewer: ```suggestion
    # get_moment returns a symbolic expression for the stable moment from which to start sampling
    # the variable, given the implicit `rv`, `size` and `param1` ... `paramN`
```

1. The `logcdf` method is not a requirement, but it's a nice plus!
1. Currently only one moment is supported in the `get_moment` method, and probably the "higher-order" one is the most useful (that is `mean` > `median` > `mode`)... You might need to truncate the moment if you are dealing with a discrete distribution.
1. When creating the `get_moment` method, we have to be careful with `size != None` and broadcast properly when some parameters that are not used in the moment may nevertheless inform about the shape of the distribution. E.g. `pm.Normal.dist(mu=0, sigma=np.arange(1, 6))` returns a moment of `[mu, mu, mu, mu, mu]`.

For a quick check that things are working you can try the following:

```python

import pymc as pm

> from pm.distributions.distribution import get_moment

Reviewer: ```suggestion
from pymc.distributions.distribution import get_moment
```

blah.eval()
# array([0.62778803, 1.95165513])

# Test the get_moment

> pm.distributions.distribution.get_moment(blah).eval()

Reviewer 2: ```suggestion
from pm.distributions.distribution import get_moment
get_moment(blah).eval()
```

blah.eval()
# array([0.62778803, 1.95165513])

# Test the get_moment
get_moment(blah).eval()

> # array([0.5, 1. ])

Reviewer 2: These would be `array([0., 0.])` with the pseudo code above no?

Author: This is what I'm getting from `get_moment()` for `blah = pm.Uniform.dist([0, 0], [1, 2])` 

I did some tests and I think this is the right output. Considering that the first argument is location and second is scale (or lower and upper), then first interval is `[0, 1]` and second interval is `[0, 2]`. With the formula `mean = 0.5 * [a, b]`, then we have `[0.5, 1.0]`. I compared with scipy implementation and got the same result. Let me now if I am mistaken here :)

```python
import pymc as pm
from pymc.distributions.distribution import get_moment
from scipy.stats import uniform 

# PyMC implementation
blah = pm.Uniform.dist([0, 0], [1, 2])
get_moment(blah).eval()
>> array([0.5, 1. ])

# spicy implementation
uniform.mean([0, 0], [1, 2])
>> array([0.5, 1. ])
```

I see that there is an error in `pre-commit / pre-commit (pull_request)` that is related with `Trim Trailing Whitespace`. Is this something I did wrong? Let me know how I can fix it.

Thank you,

Reviewer: Your comments & code here are correct. I think Ricardo referred to the example _implementation_ above.

The test case further down also assumes that `Blah == Normal`, so maybe make this example also with a normal get_moment?

Author: Thanks for the clarification! Yes, I agree it makes sense to keep it uniform by changing this example from `Uniform` to `Normal`. I will commit the changes below.

```

## 5. Adding tests for the `get_moment` method

Tests for the `get_moment` method are contained in `pymc/tests/test_distributions_moments.py`, and check if:
1. Moments return `expected` values 
1. Moments have the expected size and shape

```python

import pytest
from pymc.distributions import Blah
from pymc.initial_point import make_initial_point_fn

> def assert_moment_is_expected(model, expected):

Reviewer 2: We shouldn't copy the body of `assert_moment_is_expected`, just mention it exists, what it does and and how it should be used (which you show below)

Author: Thanks for the review! That makes sense if eventually the function `assert_moment_is_expected` changes in the test file we don't have to track the changes and update the documentation. I will work on these changes a commit them soon.

```

## 5. Adding tests for the `get_moment` method

Tests for the `get_moment` method are contained in `pymc/tests/test_distributions_moments.py`, and make use of the function `assert_moment_is_expected`
which checks if:

> 1. Moments return `expected` values

Reviewer: ```suggestion
1. Moments return the `expected` values
```
For less confusability with the statistical term _expected values_.
